### ISSUE TITLE
Caching issue

### ISSUE THREAD

**Issue by Mowee** [composer/composer#5577]
Today I ran into an caching issue. I'm using my own composer repository powered by satis. 
While creating an new release I forgot to update the "conflict" part. So my installation failed. After fixing the version conflict I tried to install all my dependencies. But Composer failed with the following message:

```
mw@laptop:~/2016.09$ bin/composer.phar update --verbose
You are running composer with xdebug enabled. This has a major impact on runtime performance. See https://getcomposer.org/xdebug
Loading composer repositories with package information
Updating dependencies (including require-dev)
Dependency resolution completed in 0.112 seconds
Analyzed 2466 packages to resolve dependencies
Analyzed 6251 rules to resolve dependencies
Dependency resolution completed in 0.003 seconds

                                                                                                                                          
  [Composer\DependencyResolver\SolverProblemsException]                                                                                   
    Problem 1                                                                                                                             
      - The requested package vendor/packagename could not be found in any version, there may be a typo in the package name.  
  Potential causes:                                                                                                                       
   - A typo in the package name                                                                                                           
   - The package is not available in a stable-enough version according to your minimum-stability setting                                  
     see <https://getcomposer.org/doc/04-schema.md#minimum-stability> for more details.                                                   
  Read <https://getcomposer.org/doc/articles/troubleshooting.md> for further common problems.                                             
                                                                                                                                          

Exception trace:
 () at phar:///var/www/2016.09/bin/composer.phar/src/Composer/DependencyResolver/Solver.php:239
 Composer\DependencyResolver\Solver->solve() at phar:///var/www/2016.09/bin/composer.phar/src/Composer/Installer.php:650
 Composer\Installer->extractDevPackages() at phar:///var/www/2016.09/bin/composer.phar/src/Composer/Installer.php:487
 Composer\Installer->doInstall() at phar:///var/www/2016.09/bin/composer.phar/src/Composer/Installer.php:216
 Composer\Installer->run() at phar:///var/www/2016.09/bin/composer.phar/src/Composer/Command/UpdateCommand.php:174
 Composer\Command\UpdateCommand->execute() at phar:///var/www/2016.09/bin/composer.phar/vendor/symfony/console/Command/Command.php:259
 Symfony\Component\Console\Command\Command->run() at phar:///var/www/2016.09/bin/composer.phar/vendor/symfony/console/Application.php:844
 Symfony\Component\Console\Application->doRunCommand() at phar:///var/www/2016.09/bin/composer.phar/vendor/symfony/console/Application.php:192
 Symfony\Component\Console\Application->doRun() at phar:///var/www/2016.09/bin/composer.phar/src/Composer/Console/Application.php:231
 Composer\Console\Application->doRun() at phar:///var/www/2016.09/bin/composer.phar/vendor/symfony/console/Application.php:123
 Symfony\Component\Console\Application->run() at phar:///var/www/2016.09/bin/composer.phar/src/Composer/Console/Application.php:104
 Composer\Console\Application->run() at phar:///var/www/2016.09/bin/composer.phar/bin/composer:43
 require() at /var/www/2016.09/bin/composer.phar:24
     
```

Deleting composers cache via "composer clear-cache" wasn't successfull at all. After a while I figured to delete all my local dependencies in projects vendor directory. Afterwards everything worked fine. 
Sounds for me something isn't working as supposed to be.

**Comment by alcohol** [comment]
Sounds to me like you are jumping to conclusions.

> While creating an new release I forgot to update the "conflict" part. So my installation failed. After fixing the version conflict I tried to install all my dependencies. 

Mind sharing how you created this release, and subsequently updated your release?

**Comment by Mowee** [comment]
Creating a release is basically just an svn copy. We have a script replacing everything with the new version. After replacing everything the new release is beeing checked in our svn. 

We also have a few plugins having the same version as our release. Here the same process applies. The script makes an svn copy and replaces everything with the new version. So far nothing special.

Now we are getting to the problem: One of our plugins had an "conflict" part in it's composer.json The conflicting package is also beeing required in the main release, but in a different version. For better understanding some example composer files

**The composer.json of our main release:**
`{
  "name": "vendor/package-main-release",
  "description": "text",
  "license": "proprietary",
  "version": "2016.07.x-dev",
  "minimum-stability": "dev",
  "prefer-stable": true,
  "require": {
    "vendor/third-package": "1.1.25",
    "vendor/package-virtual": "2016.07.*"
  },
  "repositories": [
    {
      "type": "composer",
      "url": "https://domain.url.com"
    }
  ]
}
`

**composer.json of our virtual package**
`{
  "name": "vendor/package-virtual",
  "description": "some text",
  "license": "proprietary",
  "require": {
    "vendor/package-main-release": "2016.07.*"
  },
  "conflict": {
    "vendor/third-package": "<1.1.24 | >1.1.24"
  },
  "extra": {
    "branch-alias": {
      "dev-trunk": "2016.07.x-dev"
    }
  }
}`

The problem here was that the main release required a version of the third-package that conflicted with with the virtual package. Composer failed installing the dependencies which was totally fine. After updating the composer.json of our virtual package it looked like this:

**composer.json of our virtual package afterwards**
`{
  "name": "vendor/package-virtual",
  "description": "some text",
  "license": "proprietary",
  "require": {
    "vendor/package-main-release": "2016.07.*"
  },
  "conflict": {
    "vendor/third-package": "<1.1.25 | >1.1.25"
  },
  "extra": {
    "branch-alias": {
      "dev-trunk": "2016.07.x-dev"
    }
  }
}`

But trying to install the dependencies was not working because composer wasn't capable of finding the required version.

**Comment by alcohol** [comment]
Sorry, I kind of stopped reading after the mention of "svn".

Possibly a caching issue.

Or bad configuration. You are using explicit version declarations in your composer.json files. This should be avoided like the plague.

**Comment by Mowee** [comment]
Yes I know, legacy code and so on.  svn is bad and I personally would prefer a git repository as well. But this doesn't matter because this isn't up to me. 

Yes we had a bad configuration here. Because the main package required a version that was conflicting with a version in the virtual package. Circle dependency... I already removed this to avoid future errors like this. 

But the main issue remains. As I mentioned in my first post.  I was deleting the compose cache. But only after I deleted the vendor repository the issue disappeared. So the caching issue remains.

**Comment by alcohol** [comment]
Hmm well during resolving Composer also looks at what is currently installed. If you wipe your vendor directory there is nothing that could conflict, so that obviously "fixes" things. Not sure what we can do for you here though.

**Comment by Mowee** [comment]
This doesn't makes sense for me. Because the error message states, that composer wasn't able to find this package in any version. If I have an correct version in my repository then I would expect that composer looks inside my repository as well.

**Comment by alcohol** [comment]
But you have a circular dependency. Those are known to cause issues. They are one of the very reasons that https://getcomposer.org/doc/03-cli.md#composer-root-version exists.

**Comment by alcohol** [comment]
Either way, I still don't know what you want from us. You are using a legacy setup, with bad practices, and you provide only small pieces of the complete puzzle because of "privacy" issues. I can't reproduce any of this.

**Comment by Mowee** [comment]
I resolved the circular dependency by removing the "conflict" part from my virtual package.
I cleared my local cache using composers internal clear cache function. 

Here is what I would expect: 
- A better error message. Maybe some hint that this also could be caching issue
- Clear cache should remove every cache being considered by composer without the need to manually delete the vendor directory

I agree with you: It's hard to debug this with just snippets being provided. The problem is that I can not reproduce it myself anymore. I thought my post would be helpful to avoid future confusions like mine.
